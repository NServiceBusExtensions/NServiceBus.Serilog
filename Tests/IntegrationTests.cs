using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using NServiceBus;
using NServiceBus.Logging;
using NServiceBus.Serilog;
using ObjectApproval;
using Serilog;
using Serilog.Events;
using Xunit;
using Xunit.Abstractions;

public class IntegrationTests : TestBase
{
    public static ManualResetEvent resetEvent;

    public IntegrationTests(ITestOutputHelper output) : base(output)
    {
        resetEvent = new ManualResetEvent(false);
    }

    [Fact]
    public async Task Ensure_handler()
    {
        var events = await Send(new StartHandler {Property = "TheProperty"});
        var logEvents = events.ToList();
        var logsForType = logEvents.LogForType<TheHandler>();
        ObjectApprover.VerifyWithJson(logsForType, jsonSerializerSettings: SerializerBuilder.Get());
    }

    [Fact]
    public async Task Ensure_saga()
    {
        var events = await Send(new StartSaga {Property = "TheProperty"});
        var logEvents = events.ToList();
        var sagaLogEvent = logEvents.LogForType<TheSaga>();
        ObjectApprover.VerifyWithJson(sagaLogEvent, jsonSerializerSettings: SerializerBuilder.Get());
    }

    async Task<IEnumerable<LogEvent>> Send(object message)
    {
        var logs = new ConcurrentBag<LogEvent>();
        var eventSink = new EventSink
        {
            Action = logs.Add
        };

        Log.Logger = new LoggerConfiguration()
            .MinimumLevel.Verbose()
            .WriteTo.Sink(eventSink)
            .CreateLogger();
        LogManager.Use<SerilogFactory>();

        var configuration = new EndpointConfiguration("SerilogTests");
        configuration.EnableInstallers();
        configuration.EnableSerilogTracing().EnableSagaTracing();

        configuration.SendFailedMessagesTo("error");
        configuration.UsePersistence<InMemoryPersistence>();
        configuration.UseTransport<LearningTransport>();

        var endpoint = await Endpoint.Start(configuration);
        await endpoint.SendLocal(message);
        if (!resetEvent.WaitOne(TimeSpan.FromSeconds(2)))
        {
            throw new Exception("No Set received.");
        }

        await endpoint.Stop();
        Log.CloseAndFlush();
        return logs;
    }
}