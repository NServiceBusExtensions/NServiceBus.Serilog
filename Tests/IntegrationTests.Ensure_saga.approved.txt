[
  {
    "Timestamp": "DateTimeOffset 1",
    "Level": 2,
    "MessageTemplate": {
      "Text": "Hello from {@Saga}. Message: {@Message}",
      "Tokens": {
        "$type": "MessageTemplateToken[]",
        "$values": [
          {
            "$type": "TextToken",
            "Length": 11,
            "Text": "Hello from "
          },
          {
            "$type": "PropertyToken",
            "Length": 7,
            "PropertyName": "Saga",
            "Destructuring": 2,
            "StartIndex": 11
          },
          {
            "$type": "TextToken",
            "Length": 11,
            "Text": ". Message: ",
            "StartIndex": 18
          },
          {
            "$type": "PropertyToken",
            "Length": 10,
            "PropertyName": "Message",
            "Destructuring": 2,
            "StartIndex": 29
          }
        ]
      }
    },
    "Properties": {
      "$type": "Dictionary`2",
      "Saga": {
        "$type": "ScalarValue",
        "Value": "TheSaga"
      },
      "Message": {
        "$type": "StructureValue",
        "TypeTag": "StartSaga",
        "Properties": {
          "$type": "LogEventProperty[]",
          "$values": [
            {
              "Name": "Property",
              "Value": {
                "$type": "ScalarValue",
                "Value": "TheProperty"
              }
            }
          ]
        }
      },
      "SourceContext": {
        "$type": "ScalarValue",
        "Value": "TheSaga"
      }
    }
  }
]